packages/cletus/src/operations/architect.tsx

```diff
@@ -311,3 +315,599 @@ export const type_create = operationOf<
+  Instance count: {{this.instanceCount}}
+{{/each}}
+{{else}}
+No types discovered yet in this import.
+{{/if}}
+</discoveredTypes>
+
+<constraints>
+{{#if hints}}
+Focus on discovering types that match these hints: {{hints}}
+{{/if}}
+{{#if maxTypes}}
+Maximum types to discover: {{maxTypes}}
+{{#if discoveredTypes.length}}
+Already discovered: {{discoveredTypes.length}}
+Remaining slots: {{remainingSlots}}
+{{/if}}
+{{/if}}
+</constraints>
+
+<instructions>
+Analyze the text below and extract type definitions. Use the provided tools to manage discovered types:
+- add_discovered: Add a new discovered type with fields
+- rename_discovered: Rename a discovered type
+- update_fields: Add, update, or remove fields on an existing or discovered type
+- update_discovered: Update properties of a discovered type
+- add_instances: Track instance count for a type
+- remove_discovered: Remove a discovered type (e.g., if it's too infrequent)
+
+When managing types:
+- Use estimated instance counts to prioritize important types
+- If max types is specified and reached, remove less frequent types to make room for more important ones
+- Merge similar types using update_fields and remove_discovered
+- Field names must be lowercase with no spaces
+- For existing types, only new fields can be added via update_fields
+</instructions>
+
+<text>
+{{text}}
+</text>`,
+      tools: [
+        ai.tool({
+          name: 'add_discovered',
+          description: 'Add a new discovered type definition',
+          schema: () => z.object({
+            name: z.string().describe('Type name (lowercase, no spaces)'),
+            friendlyName: z.string().describe('Display name'),
+            description: z.string().optional().describe('Type description'),
+            fields: z.array(createFieldSchema()).describe('Field definitions'),
+            instanceCount: z.number().optional().describe('Estimated number of instances found'),
+          }),
+          call: async (input) => {
+            const existing = discoveredTypes.get(input.name);
+            if (existing) {
+              return { success: false, message: `Type "${input.name}" already discovered. Use update_discovered or update_fields instead.` };
+            }
+            
+            discoveredTypes.set(input.name, {
+              name: input.name,
+              friendlyName: input.friendlyName,
+              description: input.description,
+              fields: input.fields as TypeField[],
+              instanceCount: input.instanceCount || 1,
+            });
+            
+            return { success: true, message: `Added discovered type: ${input.friendlyName}` };
+          },
+        }),
+        ai.tool({
+          name: 'rename_discovered',
+          description: 'Rename a discovered type',
+          schema: () => {
+            const discoveredTypeNamesList = Array.from(discoveredTypes.keys());
```

Comment:
This logic is duplicated a few times, pull it into function above prompt declaration